[package] use bazel-style glob matching for mock/extern

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49066 [package] use bazel-style glob matching for mock/extern**

This PR tweaks mock_module and extern_module. They are now renamed
mock and extern, and now only edit the package when a module matching
the pattern specified is required through dependency analysis.

save_extern_module and save_mock_module are added to explicitly modify
the package, but should not be needed by most users of the API unless they
are overriding require_package.

mock and extern now use bazel-style glob matching rules
(https://docs.bazel.build/versions/master/be/functions.html#glob).
i.e. `torch.**` matches `torch` and `torch.bar` but not `torchvision`.
mock and extern also now take an exclude list to filter out packages
that should not apply to the action.

Differential Revision: [D25413935](https://our.internmc.facebook.com/intern/diff/D25413935)

zdevito committed:
[package] use bazel-style glob matching for mock/extern

This PR tweaks mock_module and extern_module. They are now renamed
mock and extern, and now only edit the package when a module matching
the pattern specified is required through dependency analysis.

save_extern_module and save_mock_module are added to explicitly modify
the package, but should not be needed by most users of the API unless they
are overriding require_package.

mock and extern now use bazel-style glob matching rules
(https://docs.bazel.build/versions/master/be/functions.html#glob).
i.e. `torch.**` matches `torch` and `torch.bar` but not `torchvision`.
mock and extern also now take an exclude list to filter out packages
that should not apply to the action.

[ghstack-poisoned]

zdevito committed:
Update on "[package] use bazel-style glob matching for mock/extern"


This PR tweaks mock_module and extern_module. They are now renamed
mock and extern, and now only edit the package when a module matching
the pattern specified is required through dependency analysis.

save_extern_module and save_mock_module are added to explicitly modify
the package, but should not be needed by most users of the API unless they
are overriding require_package.

mock and extern now use bazel-style glob matching rules
(https://docs.bazel.build/versions/master/be/functions.html#glob).
i.e. `torch.**` matches `torch` and `torch.bar` but not `torchvision`.
mock and extern also now take an exclude list to filter out packages
that should not apply to the action.

Differential Revision: [D25413935](https://our.internmc.facebook.com/intern/diff/D25413935)

[ghstack-poisoned]

suo commented on torch/package/exporter.py:
```diff
@@ -303,32 +303,7 @@ def save_binary(self, package, resource, binary: bytes):
-            self.extern_module(m)
-
-    def mock_module(self, module_name: str):
+    def mock(self, include: 'GlobPattern', *, exclude: 'GlobPattern' = []):
```
lint seems fair

suo commented on torch/package/exporter.py:
```diff
@@ -343,29 +318,39 @@ def mock_module(self, module_name: str):
-            self.save_source_file('_mock', str(Path(__file__).parent / '_mock.py'), dependencies=False)
-        is_package = hasattr(self._import_module(module_name), '__path__')
-        self.save_source_string(module_name, _MOCK_IMPL, is_package, dependencies=False)
+    def extern(self, include: 'GlobPattern', *, exclude: 'GlobPattern' = []):
```
here too

suo commented on torch/package/exporter.py:
```diff
@@ -303,32 +303,7 @@ def save_binary(self, package, resource, binary: bytes):
-
-    def mock_module(self, module_name: str):
+    def mock(self, include: 'GlobPattern', *, exclude: 'GlobPattern' = []):
         """Replace the code for `module_name` in the package with a fake implementation. This module will return a fake
```
This documentation needs to be updated

suo commented on torch/package/exporter.py:
```diff
@@ -343,29 +318,39 @@ def mock_module(self, module_name: str):
 
+        Args:
+            module_name (str): e.g. "my_package.my_subpackage" the name of the external module.
+                This can also be a glob-style pattern, as described in :meth:`mock_module`
```
`mock_module` -> `mock`

suo reviewed (approved):
nice!

zdevito committed:
Update on "[package] use bazel-style glob matching for mock/extern"


This PR tweaks mock_module and extern_module. They are now renamed
mock and extern, and now only edit the package when a module matching
the pattern specified is required through dependency analysis.

save_extern_module and save_mock_module are added to explicitly modify
the package, but should not be needed by most users of the API unless they
are overriding require_package.

mock and extern now use bazel-style glob matching rules
(https://docs.bazel.build/versions/master/be/functions.html#glob).
i.e. `torch.**` matches `torch` and `torch.bar` but not `torchvision`.
mock and extern also now take an exclude list to filter out packages
that should not apply to the action.

Differential Revision: [D25413935](https://our.internmc.facebook.com/intern/diff/D25413935)

[ghstack-poisoned]

zdevito committed:
Update on "[package] use bazel-style glob matching for mock/extern"


This PR tweaks mock_module and extern_module. They are now renamed
mock and extern, and now only edit the package when a module matching
the pattern specified is required through dependency analysis.

save_extern_module and save_mock_module are added to explicitly modify
the package, but should not be needed by most users of the API unless they
are overriding require_package.

mock and extern now use bazel-style glob matching rules
(https://docs.bazel.build/versions/master/be/functions.html#glob).
i.e. `torch.**` matches `torch` and `torch.bar` but not `torchvision`.
mock and extern also now take an exclude list to filter out packages
that should not apply to the action.

Differential Revision: [D25413935](https://our.internmc.facebook.com/intern/diff/D25413935)

[ghstack-poisoned]
